### ISSUE TITLE
Full-Scratch Implementor of OAuth and OpenID Connect Talks About Findings

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#6756]
Full-Scratch Implementor of OAuth and OpenID Connect Talks About Findings<br>
<br>
https://ift.tt/2egq232<br>
<br>
Takahiko Kawasaki<br>
<br>
<div>
<section>
<div>
<div>
<div>
<div>
<div>
<div>
<div><a rel="noopener" href="https://medium.com/@darutk?source=post_page-----55015f36d1c3----------------------"><img alt="Takahiko Kawasaki" width="48" src="https://miro.medium.com/fit/c/96/96/2*iur1UduEu4gb_qVOySepLA.jpeg" height="48"></a></div>
</div>
</div>
</div>
</div>
<p>In this post, a developer who has implemented an <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc6749" target="_blank"><strong>OAuth 2.0</strong></a> <strong>and</strong> <a rel="noopener nofollow" href="http://openid.net/connect/" target="_blank"><strong>OpenID Connect</strong></a> server from scratch (me) talks about findings. Basically, consideration points for implementation are written discursively. Therefore, this is not a document for those who are looking for information like <em>&ldquo;How to set up an OAuth 2.0 and OpenID Connect server promptly&rdquo;</em>. If you are looking for such information, please visit <a rel="noopener nofollow" href="https://github.com/authlete/java-oauth-server" target="_blank"><strong>java-oauth-server</strong></a> and <a rel="noopener nofollow" href="https://github.com/authlete/java-resource-server" target="_blank"><strong>java-resource-server</strong></a> on GitHub. Using these, you can start an authorization server and a resource server, issue an access token and call a Web API with the access token, in 10 minutes with no need to set up your DB server.</p>
<h2>Bias</h2>
<p>I&rsquo;m a co-founder of <a rel="noopener nofollow" href="https://www.authlete.com/" target="_blank"><strong>Authlete, Inc.</strong></a> which is a company providing implementations of OAuth 2.0 and OpenID Connect on cloud, so this document may be affected by such a biased standpoint. Therefore, read this document with it in your mind. However, basically, I&rsquo;m going to write this post from a viewpoint of a pure engineer.</p>
<p><em>&ldquo;We want to do this and that on our corporate website. Should we implement OAuth?&rdquo;</em> &mdash; This is often asked. In essence, this question is asking what OAuth is.</p>
<p>A one-sentence answer I often use to explain OAuth is as follows.</p>
<blockquote>
<p><mark>OAuth 2.0 is a framework where a user of a service can allow a third-party application to access his/her data hosted in the service without revealing his/her credentials to the application.</mark></p>
</blockquote>
<p>The important point is <strong><em>&ldquo;Not reveal credentials to a third-party application&rdquo;</em></strong>. OAuth exists for this purpose. Once you understand this, you can judge whether you should or don&rsquo;t have to prepare an OAuth server for your company&rsquo;s service by checking whether the following conditions are satisfied.</p>
<ol>
<li>Your service manages users&rsquo; data.</li>
<li>You want third-parties to develop applications for users of your service.</li>
<li>You don&rsquo;t want to reveal credentials of users to applications developed by third-parties.</li>
</ol>
<p>Even if the conditions above are not satisfied and applications for your company&rsquo;s service are self-made ones only, it is recommended to implement an OAuth server if there is a possibility that you may want third-parties to develop applications in the future and/or if you want to follow the best practices of Web API development.</p>
<p>Still, confusion may not be resolved. It is when you want to enable users to login your website using their accounts of external services such as Facebook and Twitter. Because the term, <em>&ldquo;OAuth authentication&rdquo;</em>, is often used in this context, you may think you have to implement OAuth for your service. However, in this case, because your service is a client which uses OAuth implemented by external services, your service itself does not have to implement OAuth. To be exact, your service has to write code to use other companies&rsquo; OAuth. In other words, from a viewpoint of external services, your service has to behave as an OAuth client. However, in this use case, your service does not have to behave as an OAuth server. That is, you don&rsquo;t have to implement an OAuth server.</p>
<p>I explain the term that makes people confused &mdash; <em>&ldquo;OAuth authentication&rdquo;</em>.</p>
<p>Every explanation says <em>&ldquo;OAuth is a specification for</em> <strong><em>authorization</em></strong> <em>and not for</em> <strong><em>authentication</em></strong><em>.&rdquo;</em> It is because <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc6749" target="_blank"><strong>RFC 6749</strong></a> (The OAuth 2.0 Authorization Framework) explicitly states that authentication is &ldquo;<em>beyond the scope of this specification.&rdquo;</em> The following paragraph is an excerpt from <em>&ldquo;</em><a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc6749#section-3.1" target="_blank"><em>3.1. Authorization Endpoint</em></a><em>&rdquo;</em> in RFC 6749.</p>
<blockquote>
<p>The authorization endpoint is used to interact with the resource owner and obtain an authorization grant. The authorization server MUST first verify the identity of the resource owner. The way in which the authorization server authenticates the resource owner (e.g., username and password login, session cookies) is <strong>beyond the scope of this specification</strong>.</p>
</blockquote>
<p>Nevertheless, the term &ldquo;OAuth authentication&rdquo; is flooding and making people confused. This confusion is observed not only among business-side people but also among engineers. For example, questions like &ldquo;<a rel="noopener nofollow" href="http://stackoverflow.com/q/33702826/1174054" target="_blank">OAuth Authorization vs Authentication</a>&rdquo; are sometimes posted to Stack Overflow (my answer to the question is <a rel="noopener nofollow" href="http://stackoverflow.com/a/33704657/1174054" target="_blank">this</a>).</p>
<p>Information dealt by the terms, authentication and authorization (in the context of OAuth), can be described as shown below.</p>
<ul>
<li>
<strong>Authentication</strong> &mdash; Who one is.</li>
<li>
<strong>Authorization</strong> &mdash; Who grants what permissions to whom.</li>
</ul>
<p>Authentication is a simple concept. In other words, it is confirmation of identity. The most prevailing way to identify a person at a website is to request the person to present a pair of ID and password, but there are other ways such as biometric authentication using fingerprint or iris, one-time password, random number table and so on. In any case, whatever way is used, authentication is a process to identify who one is. Using developer words, it can be expressed like <strong><em>&ldquo;Authentication is a process to identify the unique identifier of a user.&rdquo;</em></strong></p>
<p>On the other hand, authorization is complicated because three elements, namely, <em>&ldquo;who&rdquo;</em>, <em>&ldquo;what permissions&rdquo;</em> and <em>&ldquo;to whom&rdquo;</em>, are involved. In addition, what makes it confusing is that among the three elements, the process to identify <em>&ldquo;who&rdquo;</em> is authentication. In other words, the fact that <strong>authorization process includes authentication process as a part</strong> is making things confusing.</p>
<p>If the three elements should be replaced with words used by developers, <em>&ldquo;who&rdquo;</em> can be replaced with <em>&ldquo;user&rdquo;</em>, and <em>&ldquo;to whom&rdquo;</em> with <em>&ldquo;client application&rdquo;</em>. As a result, authorization in the context of OAuth can be said to be the process where a user grants permissions to a client application.</p>
<p>The figure below depicts the concept explained so far.</p>
<figure>
<div>
<div>
<div>
<div>
<div><img width="687" src="https://miro.medium.com/max/60/1*r4AcU1CSDokwE0a03P9kZQ.png?q=20" height="488"></div>
<img src="https://miro.medium.com/max/1374/1*r4AcU1CSDokwE0a03P9kZQ.png" width="687" height="488">
</div>
</div>
</div>
</div>
</figure>
<p>This figure illustrates which parts in an authorization page (a page where a user grants permissions to a client application) are for authentication and for authorization. The difference between authentication and authorization is clear.</p>
<p>Now, it&rsquo;s time to talk about <em>&ldquo;OAuth authentication&rdquo;</em>.</p>
<p>Because authorization process includes authentication process as a part, being authorized means being authenticated. So, some people began to use OAuth for authentication. This is <em>&ldquo;OAuth authentication&rdquo;</em> and it has prevailed rapidly because of some merits such as &ldquo;that the task to manage user credentials can be delegated to external services&rdquo; and &ldquo;that the hurdle for new users to start using the service becomes lower because the user registration process can be omitted.&rdquo;</p>
<p>OpenID guys held a grudge against the situation. &mdash;Sorry, I don&rsquo;t know whether they actually felt in that way, but at least I can imagine that they felt OAuth authentication was far from the level of specifications that they had defined by then such as OpenID 2.0 and SAML. However, it was undeniable fact that their specifications had not prevailed very much and developers in the world had chosen the easiness of OAuth authentication. Therefore, they have defined a new specification for authentication, <a rel="noopener nofollow" href="http://openid.net/connect/" target="_blank"><strong>OpenID Connect</strong></a>, on top of OAuth. <a rel="noopener nofollow" href="http://openid.net/connect/faq/" target="_blank">OpenID Connect FAQ</a> depicts the relationship as an equation like below.</p>
<pre><span>(Identity, Authentication) + OAuth 2.0 = OpenID Connect</span></pre>
<p>Thanks to this, authentication by OpenID Connect can be executed at the same time during the authorization process by OAuth.</p>
<p>As major players in the industry have been working on specification creation and proactively implementing it (<a rel="noopener nofollow" href="http://openid.net/connect/faq/" target="_blank">FAQ</a>), OpenID Connect will surely prevail. As a result, OAuth authentication libraries such as <a rel="noopener nofollow" href="https://github.com/intridea/omniauth" target="_blank"><strong>OmniAuth</strong></a> will gradually finish their roles.</p>
<p>However, people will surely be made more confused because OpenID Connect which is for authentication has been built on top of OAuth which is for authorization. It is difficult to explain, especially in my case because Authlete focuses on authorization and does not do anything for authentication although it supports OpenID Connect<em>.</em> I always have to explain the difference between authentication and authorization before starting to explain the product itself to customers.</p>
<p>Regarding the problem of OAuth authentication, please read an article <em>&ldquo;</em><a rel="noopener nofollow" href="http://www.thread-safe.com/2012/01/problem-with-oauth-for-authentication.html" target="_blank"><em>The problem with OAuth for Authentication</em></a><em>&rdquo;</em> by Mr. John Bradley. In the article he says <em>&ldquo;This is a security hole that you can drive a house through.&rdquo;</em></p>
<figure>
<div>
<div>
<div>
<div>
<div><img width="2048" src="https://miro.medium.com/max/60/1*Hs65d5VydD2lYPjrM6tY7g.jpeg?q=20" height="1124"></div>
<img src="https://miro.medium.com/max/4096/1*Hs65d5VydD2lYPjrM6tY7g.jpeg" width="2048" height="1124">
</div>
</div>
</div>
</div>
<figcaption>&ldquo;Say OAuth is an Authentication standard again.&rdquo; by Mr. Nat Sakimura and Mr. John Bradley. (from <a rel="noopener nofollow" href="https://twitter.com/ve7jtb/status/740650395735871488" target="_blank">https://twitter.com/ve7jtb/status/740650395735871488</a>)</figcaption>
</figure>
<p>All the contents so far are, however, just a preamble of this post. Technical contents for developers start from here. The first topic is the relationship between OAuth 2.0 and OpenID Connect.</p>
<p>It was after I had finished implementing <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc6749" target="_blank">RFC 6749</a> (The OAuth 2.0 Authorization Framework) that I noticed the existence of <a rel="noopener nofollow" href="http://openid.net/connect/" target="_blank">OpenID Connect</a>. As I gather information about OpenID Connect, I thought I should implement the feature and so read <a rel="noopener nofollow" href="http://openid.net/specs/openid-connect-core-1_0.html" target="_blank">OpenID Connect Core 1.0</a> and other related specifications. After reading, the conclusion I had reached was <strong><em>&ldquo;All should be rewritten from scratch.&rdquo;</em></strong></p>
<p>The OpenID Connect website says <em>&ldquo;OpenID Connect 1.0 is a simple identity layer on top of the OAuth 2.0 protocol.&rdquo;</em> and this gives an impression that OpenID Connect can be implemented easily and seamlessly on top of an existing OAuth 2.0 implementation. However, the truth is utterly different. IMHO, OpenID Connect is virtually OAuth 3.0.</p>
<p>There are many specifications related to OpenID Connect and they are so puzzling that it is hard to decipher them. I almost went crazy and had to read them three times before I could manage to grasp the entire picture. Compared to OpenID Connect specifications, RFC 6749 can be said to be easy.</p>
<p>Especially, what conflicts with an existing implementation is the way to process the request parameter, <code>response_type</code>. It is sure that RFC 6749 states the request parameter may take multiple values, but it is just a possibility in the future. If we read RFC 6749 straightforwardly, <code>response_type</code> is either <code>code</code> or <code>token</code>. It is almost impossible to imagine that these two are set at the same time. It is because the parameter is used to determine the flow to process a request from a client application. To be concrete, the <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc6749#section-1.3.1" target="_blank">authorization code flow</a> is used when the value of <code>response_type</code> is <code>code</code>, and the <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc6749#section-1.3.2" target="_blank">implicit flow</a> is used when the value is <code>token</code>. Who can imagine these flows are mixed? Even if one could imagine it, how should we resolve conflicts that exist between the flows? For example, the authorization code flow requires that response parameters be embedded in the <em>query</em> part of a redirect URI (<a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc6749#section-4.1.2" target="_blank">4.1.2. Authorization Response</a>) while the implicit flow requires that response parameters be embedded in the <em>fragment</em> part (<a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc6749#section-4.2.2" target="_blank">4.2.2. Access Token Response</a>), and these requirements cannot be satisfied simultaneously.</p>
<p>However, OpenID Connect has added <code>id_token</code> as a new value for <code>response_type</code> and explicitly allowed any combination of <code>code</code>, <code>token</code> and <code>id_token</code> as a value of <code>response_type</code>. Furthermore, <code>none</code> is added, too. Details are described in &ldquo;<a rel="noopener nofollow" href="http://openid.net/specs/openid-connect-core-1_0.html#Authentication" target="_blank">3. Authentication</a>&rdquo; of <a rel="noopener nofollow" href="http://openid.net/specs/openid-connect-core-1_0.html" target="_blank">OpenID Connect Core 1.0</a> and in <a rel="noopener nofollow" href="http://openid.net/specs/oauth-v2-multiple-response-types-1_0.html" target="_blank">OAuth 2.0 Multiple Response Type Encoding Practices</a>.</p>
<p>It requires major changes to modify existing code written on the assumption of the either-or choice so that it can handle any combination of possible values and mixed flows. Therefore, implementors of OAuth library should write it with OpenID Connect in their head from the beginning if there is a possibility to support OpenID Connect in the future. In other words, existing OAuth libraries cannot support OpenID Connect without major modifications.</p>
<p><a rel="noopener nofollow" href="https://github.com/spring-projects/spring-security-oauth" target="_blank"><strong>Spring Security OAuth</strong></a>, for example. This library has not supported OpenID Connect yet (as of June 2016). For the library to support OpenID Connect, to begin with, the request parameter <code>response_type</code> has to be able to take other values than <code>code</code> and <code>token</code>. A request for it is listed as <em>&ldquo;</em><a rel="noopener nofollow" href="https://github.com/spring-projects/spring-security-oauth/issues/619" target="_blank"><em>Issue #619 Handling additional response_types</em></a><em>&rdquo;</em>, but it has not been resolved yet and the last comment of the thread is <em>&ldquo;Any comments are more than welcome as this turned out to be (as I predicted) a large refactor exercise.&rdquo;</em> I read some relevant source files and found it true that supporting OpenID Connect would require big changes. Unless some companies back up the project financially, I&rsquo;m afraid it will take long time for the project to support OpenID Connect.</p>
<p>In passing, I&rsquo;d like to mention <a rel="noopener nofollow" href="https://oltu.apache.org/" target="_blank"><strong>Apache Oltu</strong></a>, too. The project claims that it supports OpenID Connect, but my guess is that the initial implementation supported OAuth 2.0 only and OpenID Connect support was added in a later phase. The reason I think so is that the package for OAuth 2.0 ( <code><a rel="noopener nofollow" href="https://oltu.apache.org/apidocs/oauth2/reference/packages.html" target="_blank">org.apache.oltu.oauth2</a></code>) and that for OpenID Connect ( <code><a rel="noopener nofollow" href="https://oltu.apache.org/apidocs/openid-connect/index.html" target="_blank">org.apache.oltu.openidconnect</a></code>) are isolated. However, this kind of approach distorts the architecture. For example, it is inappropriate that <code><a rel="noopener nofollow" href="https://oltu.apache.org/apidocs/openid-connect/reference/org/apache/oltu/openidconnect/client/response/OpenIdConnectResponse.html" target="_blank">OpenIdConnectResponse</a></code> class is a descendant of <code><a rel="noopener nofollow" href="https://oltu.apache.org/apidocs/oauth2/reference/org/apache/oltu/oauth2/client/response/OAuthAccessTokenResponse.html" target="_blank">OAuthAccessTokenResponse</a></code> because a response containing an ID token does not necessarily contain an access token. Other examples are existence of a class named <code><a rel="noopener nofollow" href="https://oltu.apache.org/apidocs/oauth2/reference/org/apache/oltu/oauth2/client/request/OAuthClientRequest.AuthenticationRequestBuilder.html" target="_blank">OAuthClientRequest.AuthenticationRequestBuilder</a></code> (not &ldquo;Authorization&rdquo; but &ldquo;Authentication&rdquo; for some reasons) and existence of a GitHub-specific class <code><a rel="noopener nofollow" href="https://oltu.apache.org/apidocs/oauth2/reference/org/apache/oltu/oauth2/client/response/GitHubTokenResponse.html" target="_blank">GitHubTokenResponse</a></code>. The architecture of Apache Oltu brings up questions at least to me. I don&rsquo;t know the details about the project, but in my personal opinion, it is destined to shrink.</p>
<p>As written explicitly in <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc6749#section-2" target="_blank">2. Client Registration</a> of RFC 6749, a client application has to be registered into the target authorization server in advance before it makes an authorization request. Therefore, in a typical case, an implementor of an authorization server defines a database table to store information about client applications.</p>
<p>To decide what columns the table should have, an implementor lists up items by reading the specification. For example, reading RFC 6749 will make you realize that at least the following items are necessary.</p>
<ol>
<li>Client ID</li>
<li>Client Secret</li>
<li>Client Type</li>
<li>Redirect URIs</li>
</ol>
<p>In addition to these, an implementor may add more attributes. For instance, &ldquo;application name&rdquo;.</p>
<p>There are not so many attributes of a client application even if you rifle through RFC 6749, so the number of columns of the database table to store attributes of client applications won&rsquo;t become big &mdash; such good old days have ended due to the emergence of OpenID Connect. Many attributes that a client application should have are listed in <a rel="noopener nofollow" href="http://openid.net/specs/openid-connect-registration-1_0.html#ClientMetadata" target="_blank">2. Client Metadata</a> of <a rel="noopener nofollow" href="http://openid.net/specs/openid-connect-registration-1_0.html" target="_blank">OpenID Connect Dynamic Client Registration 1.0</a>. Below is the list.</p>
<ol>
<li>
<code>redirect_uris</code> &mdash; Redirect URI values used by the Client.</li>
<li>
<code>response_types</code> &mdash; <code>response_type</code> values that the Client is declaring that it will restrict itself to using.</li>
<li>
<code>grant_types</code> &mdash; Grant Types that the Client is declaring that it will restrict itself to using.</li>
<li>
<code>application_type</code> &mdash; Kind of the application.</li>
<li>
<code>contacts</code> &mdash; e-mail addresses of people responsible for this Client.</li>
<li>
<code>client_name</code> &mdash; Name of the Client to be presented to the End-User.</li>
<li>
<code>logo_uri</code> &mdash; URL that references a logo for the Client application.</li>
<li>
<code>client_uri</code> &mdash; URL of the home page of the Client.</li>
<li>
<code>policy_uri</code>&mdash; URL that the Relying Party Client provides to the End-User to read about how the profile data will be used.</li>
<li>
<code>tos_uri</code>&mdash; URL that the Relying Party Client provides to the End-User to read about the Relying Party&rsquo;s terms of service.</li>
<li>
<code>jwks_uri</code>&mdash; URL for the Client&rsquo;s JSON Web Key Set document.</li>
<li>
<code>jwks</code> &mdash; Client&rsquo;s JSON Web Key Set document, passed by value.</li>
<li>
<code>sector_identifier_uri</code> &mdash; URL using the https scheme to be used in calculating Pseudonymous Identifiers by the OP.</li>
<li>
<code>subject_type</code> &mdash; <code>subject_type</code> requested for responses to this Client.</li>
<li>
<code>id_token_signed_response_alg</code> &mdash; JWS <code>alg</code> algorithm required for signing the ID Token issued to this Client.</li>
<li>
<code>id_token_encrypted_response_alg</code> &mdash; JWE <code>alg</code> algorithm required for encrypting the ID Token issued to this Client.</li>
<li>
<code>id_token_encrypted_response_enc</code>&mdash; JWE <code>enc</code> algorithm required for encrypting the ID Token issued to this Client.</li>
<li>
<code>userinfo_signed_response_alg</code>&mdash; JWS <code>alg</code> algorithm required for signing UserInfo Responses.</li>
<li>
<code>userinfo_encrypted_response_alg</code> &mdash; JWE <code>alg</code> algorithm required for encrypting UserInfo Responses.</li>
<li>
<code>userinfo_encrypted_response_enc</code> &mdash; JWE <code>enc</code> algorithm required for encrypting UserInfo Responses.</li>
<li>
<code>request_object_signing_response_alg</code> &mdash; JWS <code>alg</code> algorithm that must be used for signing Request Objects sent to the OP.</li>
<li>
<code>request_object_encryption_alg</code> &mdash; JWE <code>alg</code> algorithm that RP is declaring that it may use for encrypting Request Object sent to the OP.</li>
<li>
<code>request_object_encryption_enc</code> &mdash; JWE <code>enc</code> algorithm the RP is declaring that it may use for encrypting Request Objects sent to the OP.</li>
<li>
<code>token_endpoint_auth_method</code>&mdash; Requested Client Authentication method for the Token Endpoint.</li>
<li>
<code>token_endpoint_auth_signing_alg</code> &mdash; JWS <code>alg</code> algorithm that must be used for signing the JWT used to authenticate the Client at the Token Endpoint for the <code>private_key_jwt</code> and <code>client_secret_jwt</code> authentication methods.</li>
<li>
<code>default_max_age</code> &mdash; Default Maximum Authentication Age.</li>
<li>
<code>require_auth_time</code> &mdash; Boolean value specifying whether the <code>auth_time</code> Claim in the ID Token is required.</li>
<li>
<code>default_acr_values</code> &mdash; Default requested Authentication Context Class Reference values.</li>
<li>
<code>initiate_login_uri</code> &mdash; URI using the https scheme that a third party can use to initiate a login by the RP.</li>
<li>
<code>request_uris</code> &mdash; <code>request_uri</code> values that are pre-registered by the RP for use at the OP.</li>
</ol>
<p>Therefore, a database table for client applications should be able to store these pieces of information. In addition, it should be noted that some attributes (such as <code>client_name</code>, <code>tos_uri</code>, <code>policy_uri</code>, <code>logo_uri</code> and <code>client_uri</code>) are allowed to be localized (<a rel="noopener nofollow" href="http://openid.net/specs/openid-connect-registration-1_0.html#LanguagesAndScripts" target="_blank">2.1. Metadata Languages and Scripts</a>). Additional consideration for database table design will be required to store localized attribute values.</p>
<p>Subsections hereafter are my personal opinions about client application attributes.</p>
<h2>6.1. Client Type</h2>
<p>I&rsquo;m afraid it is a kind of a mistake in defining the specification that <a rel="noopener nofollow" href="http://openid.net/specs/openid-connect-registration-1_0.html#ClientMetadata" target="_blank">2. Client Metadata</a> of <a rel="noopener nofollow" href="http://openid.net/specs/openid-connect-registration-1_0.html" target="_blank">OpenID Connect Dynamic Client Registration 1.0</a> does not contain &ldquo;client type&rdquo;. The reason I think so is that the difference between the two client types, &ldquo;<strong>confidential</strong>&rdquo; and &ldquo;<strong>public</strong>&rdquo; (which are defined in <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc6749#section-2.1" target="_blank">2.1. Client Types</a> of RFC 6749), must be taken into consideration when we implement an authorization server. As a matter of fact, &ldquo;client type&rdquo; is listed as an example of client properties to be registered in <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc6749#section-2" target="_blank">2. Client Registration</a> of RFC 6749 as follows.</p>
<blockquote>
<p>&hellip;registration can rely on other means for establishing trust and obtaining the required client properties (e.g., redirection URI, <strong>client type</strong>).</p>
</blockquote>
<p>If this is not a mistake, there must be a consensus on the client type of client applications that are registered by Dynamic Client Registration. However, I could not find such information in relevant specifications.</p>
<p>In any case, I think that a column for client type should exist when defining a database table for client applications.</p>
<p>You can find some discussion about this in <a rel="noopener nofollow" href="https://bitbucket.org/openid/connect/issues/991" target="_blank">Issue 991</a>.</p>
<h2>6.2. Application Type</h2>
<p>According to the specification, <code>application_type</code> is an optional attribute. Pre-defined values for <code>application_type</code> are <code>native</code> and <code>web</code>. If omitted, <code>web</code> is used as the default value.</p>
<p>If the default value is used when omitted, the natural consequence is that the application type of a client application must be either <code>native</code> and <code>web</code>. So, you may feel like adding <code>NOT NULL</code> to the column for <code>application_type</code>. However, the implementation of Authlete dare not add <code>NOT NULL</code> and allows <code>NULL</code>.</p>
<p>The reason is that I&rsquo;m not sure that the restrictions on redirect URI values imposed by <code>application_type</code>, which are defined in OpenID Connect Dynamic Client Registration 1.0 as follows, should be applied to every OAuth 2.0 client.</p>
<blockquote>
<p>Web Clients using the OAuth Implicit Grant Type MUST only register URLs using the https scheme as redirect_uris; they MUST NOT use localhost as the hostname. Native Clients MUST only register redirect_uris using custom URI schemes or URLs using the http: scheme with localhost as the hostname.</p>
</blockquote>
<p>2 years ago, I posted a question &ldquo;<a rel="noopener nofollow" href="http://stackoverflow.com/q/23557801/1174054" target="_blank">Does Application Type (OpenID Connect) correspond to Client Type (OAuth 2.0)?</a>&rdquo; to Stack Overflow, but I could not get any answer. So I investigated and <a rel="noopener nofollow" href="http://stackoverflow.com/a/23583796/1174054" target="_blank">answered by myself</a>. Please see it if interested.</p>
<h2>6.3. Client Secret</h2>
<p>How long should the length of client secret be?</p>
<p>For example, <a rel="noopener nofollow" href="https://backstage.forgerock.com/#!/docs/openam/12.0.0/admin-guide" target="_blank">OpenAM Administration Guide</a> uses <code>password</code> as an example of client secret value. Below is a screenshot of <a rel="noopener nofollow" href="https://backstage.forgerock.com/#%21/docs/openam/12.0.0/admin-guide/chap-oauth2#oauth2-client-plus-authz" target="_blank">12.4.1. Configuring OpenAM as Authorization Server &amp; Client</a>.</p>
<figure>
<div>
<div>
<div>
<div><img width="671" src="https://miro.medium.com/max/60/1*jp-y7qtwfD-8L_zXv9HeKQ.png?q=20" height="127"></div>
<img src="https://miro.medium.com/max/1342/1*jp-y7qtwfD-8L_zXv9HeKQ.png" width="671" height="127">
</div>
</div>
</div>
</figure>
<p>It seems OpenAM allows users to use a short string as a client secret.</p>
<p>On the other hand, in the implementation of Authlete, client secrets are generated automatically and become long like the following.</p>
<pre><span>GBAyfVL7YWtP6gudLIjbRZV_N0dW4f3xETiIxqtokEAZ6FAsBtgyIq0MpU1uQ7J08xOTO2zwP0OuO3pMVAUTid</span></pre>
<p>The reason for this length is that I wanted to support 512 bits for symmetric signature and encryption algorithms. For instance, I wanted to support <code>HS512</code> as a signature algorithm for <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc7515" target="_blank"><strong>JWS</strong></a>. Because client secrets have to have entropy of 512 bits or more to support <code>HS512</code>, the length of the example above is 86, which is a result of encoding 512-bit data using base64url.</p>
<p>Regarding entropy for symmetric signature and encryption algorithms, <a rel="noopener nofollow" href="http://openid.net/specs/openid-connect-core-1_0.html#SymmetricKeyEntropy" target="_blank">16.19 Symmetric Key Entropy</a> in OpenID Connect Core 1.0 states as follows.</p>
<blockquote>
<p>In Section 10.1 and Section 10.2, keys are derived from the client_secret value. Thus, when used with symmetric signing or encryption operations, client_secret values MUST contain sufficient entropy to generate cryptographically strong keys. Also, <strong>client_secret values MUST also contain at least the minimum of number of octets required for MAC keys for the particular algorithm used.</strong> So for instance, for HS256, the client_secret value MUST contain at least 32 octets (and almost certainly SHOULD contain more, since client_secret values are likely to use a restricted alphabet).</p>
</blockquote>
<p>And, <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc7518#section-3.1" target="_blank">3.1. &ldquo;alg&rdquo; (Algorithm) Header Parameter Values for JWS</a> in <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc7518" target="_blank">RFC 7518</a> (JSON Web Algorithms) states that <code>HS256</code> (HMAC using SHA-256) must be supported as a signature algorithm for JWS. As a logical consequence, any implementation claiming compliance with OpenID Connect is required to generate client secrets with entropy of 256 bits or more.</p>
<h2>6.4. Signature Algorithm</h2>
<p><code>id_token_signed_response_alg</code> is listed in &ldquo;2. Client Metadata&rdquo; of OpenID Connect Dynamic Client Registration 1.0. It denotes the algorithm that a client application requires the authorization server to use as a signature algorithm for ID tokens. Valid values are listed in RFC 7518 as mentioned above, and it should be noted that <code>none</code> is not allowed. If the value of <code>id_token_signed_response_alg</code> is omitted on registration, <code>RS256</code> is used.</p>
<p><code>userinfo_signed_response_alg</code> is also a signature algorithm that a client application requires the authorization server to use. This algorithm is used to sign information returned from <strong>UserInfo Endpoint</strong> (OpenID Connect Core 1.0, <a rel="noopener nofollow" href="http://openid.net/specs/openid-connect-core-1_0.html#UserInfo" target="_blank">5.3. UserInfo Endpoint</a>). <code>none</code> is allowed, and in that case, the endpoint returns <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc7515#appendix-A.5" target="_blank">Unsecured JWS</a> (= JWS without signature). In addition, <code>userinfo_signed_response_alg</code> can remain unspecified, and in this case, the endpoint returns information in plain JSON format.</p>
<p>Based on the information above, it will be okay to add <code>NOT NULL</code> to the column for <code>id_token_signed_response_alg</code>, but it will be better not to add <code>NOT NULL</code> to the column for <code>userinfo_signed_response_alg</code> and to allow <code>NULL</code> to enable the column to express the meaning &ldquo;unspecified&rdquo;. Of course, an implementor can choose to define a special value which means &ldquo;unspecified&rdquo; and store the value instead of <code>NULL</code>. If this design is adopted, <code>NOT NULL</code> can be added to the column for <code>userinfo_signed_response_alg</code>. It&rsquo;s up to implementors.</p>
<p>From among <a rel="noopener nofollow" href="http://openid.net/developers/libraries/#jwt" target="_blank">existing JWT libraries</a>, I chose <a rel="noopener nofollow" href="https://bitbucket.org/connect2id/nimbus-jose-jwt/wiki/Home" target="_blank"><strong>Nimbus JOSE+JWT</strong></a>. However, while using it, I noticed that the library did not allow signing with <code>none</code> algorithm. In other words, the library cannot generate Unsecured JWS. If this restriction exists, <code>userinfo_signed_response_alg=none</code> cannot be supported. So, I sent a <a rel="noopener nofollow" href="https://bitbucket.org/connect2id/nimbus-jose-jwt/pull-requests/19/supported-jws-alg-none/diff" target="_blank">pull request</a> to enable signing with <code>none</code>. However, it was rejected. The reason was that it was the library&rsquo;s policy not to allow signing with <code>none</code>. According to the comment, <em>&ldquo; There is significant security risk of confusing developers that alg=none objects can be signed or verified in the same way as truly protected JWS objects&rdquo;</em>, and I was advised to treat the case of <code>alg=none</code> as a special one.</p>
<p>Well, there are various opinions, but mine is that it is not a good approach to save naive developers by restricting features. As a matter of fact, Nimbus JOSE+JWT library cannot generate <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc7515#appendix-A.5" target="_blank">Unsecured JWS</a> which is listed in <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc7515" target="_blank">RFC 7515</a> as an example although there was at least one developer (me) who understood what Unsecured JWS is and still wanted to generate one. Therefore, I had to generated Unsecured JWS manually. (But, it should be fairly noted that I think Nimbus JOSE+JWT is a good library because it supports all the algorithms listed in <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc7518" target="_blank">RFC 7518</a>.)</p>
<p>This is off topic, but a certain issue was created for <a rel="noopener nofollow" href="https://github.com/TakahikoKawasaki/nv-websocket-client" target="_blank"><strong>nv-websocket-client</strong></a> (<a rel="noopener nofollow" href="http://darutk-oboegaki.blogspot.jp/2015/04/websocket-java-se-15-android.html" target="_blank">info in Japanese</a>) which is a WebSocket client library for Java I open to the public on GitHub. The issue was a proposal of feature improvement suggesting that the library have a mechanism to warn when a developer calls both <code>setSSLContext()</code> method and <code>setSSLSocketFactory()</code> method. This proposal was made because the reporter was troubled with unintended behaviors as he had called the both methods improperly. My answer was that it is explicitly written in the <a rel="noopener nofollow" href="http://takahikokawasaki.github.io/nv-websocket-client/" target="_blank">JavaDoc</a> about which setting takes precedence when the both methods are called and that such a meddling check would make <code>WebSocketFactory</code> class awkward to use. Then, the reaction was <em>&ldquo;It was my fault for not reading the documentation in details first before calling both methods. But how many other developers do you think will read the documentation in details first before making the same mistake?&rdquo;</em></p>
<p>Oh, it&rsquo;s just a well-deserved punishment if a developer wastes time on self-made bugs due to the reason he/she has not read documents&hellip;</p>
<p>Trials to help those who don&rsquo;t read documents would be endless. Even if a library prevents signing by <code>alg=none</code>, such engineers would not hesitate to include private keys in a <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc7517" target="_blank"><strong>JWK Set</strong></a> which is published through a JWK Set endpoint of an authorization server. Why? Do you think that those who don&rsquo;t read documents can notice the existence of <code>toPublicJWKSet()</code> method of <code>JWKSet</code> class (in Nimbus JOSE+JWT library) and understand the meaning of the method? Probably, they would naively say, <em>&ldquo;Yes, I could create an instance of JWKSet class. Let&rsquo;s publish it! I&rsquo;ve finished implementing the JWK Set endpoint!&rdquo;</em></p>
<p>Engineers who don&rsquo;t refer to primary sources such as RFCs cannot notice <a rel="noopener nofollow" href="http://stackoverflow.com/a/26791665/1174054" target="_blank">errors in answers</a> they find and believe the answers without doubt. However, <a rel="noopener nofollow" href="http://stackoverflow.com/a/34554243/1174054" target="_blank">an engineer must not avoid reading RFCs to become a true engineer</a>.</p>
<blockquote>
<p>To become a true engineer, don&rsquo;t avoid reading RFCs. Only searching technical blogs and Stack Overflow for answers will never bring you to the right place.</p>
</blockquote>
<h2>6.5. Client Application Developer</h2>
<p>Some open-source authorization servers provide a mechanism to enable dynamic registration of client applications such an HTML form (<a rel="noopener nofollow" href="https://forgerock.org/openam/" target="_blank"><strong>OpenAM</strong></a> by <a rel="noopener nofollow" href="https://www.forgerock.com/" target="_blank">ForgeRock</a>) and Web APIs (<a rel="noopener nofollow" href="https://github.com/mitreid-connect/" target="_blank"><strong>MITREid Connect</strong></a> by <a rel="noopener nofollow" href="http://www.mitre.org/" target="_blank">MITRE</a>). But, it seems only administrators of authorization servers can register client applications. However, an ideal approach will be to create something similar to T<a rel="noopener nofollow" href="https://apps.twitter.com/" target="_blank">witter&rsquo;s Application Management console</a>, let developers login there, and provide an environment in which each developer can register and manage his/her own client applications. To achieve this, a database table for client applications should have a column which holds developers&rsquo; unique identifiers.</p>
<p>It is often forgotten because implementing an authorization server is cumbersome in itself, but <strong>a mechanism to manage client applications also needs to be provided in order to open Web APIs to the public</strong>. If intended users of your Web APIs are limited to closed groups, the administrator of your authorization server may be able to register a client application every time he/she is requested to. As a matter of fact, there is a company whose administrator types SQL statements manually for each registration request. However, if you want to open Web APIs to the general public, such operations won&rsquo;t work and you will realize that you have to provide a decent management console for client applications. If you succeed in securing budget for development of an authorization server and Web APIs but forget to secure budget for a management console for client applications, it will result in &ldquo;Has implemented Web APIs but cannot open them to the public&rdquo;.</p>
<p>As an example of such management consoles, Authlete provides <a rel="noopener nofollow" href="https://www.authlete.com/documents/cd_console" target="_blank"><strong>Developer Console</strong></a> for use cases described above. Authlete itself does NOT manage developer accounts, but by a mechanism named &ldquo;<a rel="noopener nofollow" href="https://www.authlete.com/documents/definitive_guide/developer_authentication_callback" target="_blank">developer authentication callback</a>&rdquo;, developers whose accounts are managed by Authlete&rsquo;s customers can use the developer console. Therefore, Authlete customers don&rsquo;t have to develop a management console for client applications.</p>
<h2>7.1. Access Token Representation</h2>
<p>How should an access token be represented? There are two major ways.</p>
<ol>
<li>As a meaningless random string. Information associated with an access token is stored in a database table behind an authorization server.</li>
<li>As a self-contained string which is a result of encoding access token information by <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc4648#section-5" target="_blank">base64url</a> or something similar.</li>
</ol>
<p>A choice between these ways will lead to consequent differences as described in the following table.</p>
<figure>
<div>
<div>
<div>
<div>
<div><img width="1095" src="https://miro.medium.com/max/60/1*TAD4aF48QhUxV53rLG_vAw.png?q=20" height="769"></div>
<img width="1095" height="769">
</div>
</div>
</div>
</div>
</figure>
<p>If access tokens are random strings, it is needed to inquire of an authorization server every time in order to get information about access tokens. By contrast, if access tokens themselves contain information, there is no need to inquire of an authorization server. This makes the self-contained style sound better, but because an inquiry to an authorization server has to be made to check whether an access token has been revoked or not, even if the self-contained style is adopted, in any case, network communication is required every time an access token is presented by a client application.</p>
<p>What is cumbersome in the self-contained style is that we have to add a record denoting &ldquo;revoked&rdquo; every time access token revocation is requested, and such a record must be kept until the access token expires. Otherwise, if the record were deleted, the revoked access token would be resurrected and become valid again (if the original expiration date has not been reached yet).</p>
<p>By contrast, in the case of the random-string style, access token revocation can be achieved simply by deleting the access token record itself. Therefore, there is no way for a revoked access token to be resurrected due to any accident. In addition, the negative effect &ldquo;revocation increases records&rdquo; observed in the self-contained style won&rsquo;t happen.</p>
<p>To enable access token revocation, unique identifiers must be assigned to access tokens even in the case of self-contained style. Otherwise, it is impossible to tell which access token has been revoked. To put it the other way around, an authorization server which adopts the self-contained style but does not assign unique identifiers to access tokens is an authorization server which cannot revoke access tokens. It could be one of implementation policies, but such an authorization server should not issue long-lived access tokens and should not issue refresh tokens.</p>
<p><em>&ldquo;Authorization server that cannot revoke access tokens?!&rdquo;</em>, you might wonder. But, such an authorization actually exists. A certain global big system integrator bought a company and was developing an authorization server using the product of the acquired company, but in a later phase, the system integrator and its customer noticed that the authorization server could not revoke access tokens. When I heard the story, I guessed that the authorization server issues access tokens of self-contained style without unique identifiers.</p>
<p>The self-contained style is seemingly good because there are some merits such as &ldquo;no need to inquire of an authorization server to extract information of access tokens&rdquo; and &ldquo;no need to maintain access token records on the authorization server side&rdquo;, but when you consider access token revocation into account, there is room for discussion.</p>
<h2>7.2. Access Token Deletion</h2>
<p>To prevent a database from growing infinitely, expired access tokens should be deleted from the database periodically.</p>
<p>Client applications which request an authorization server to issue access tokens unnecessarily are trouble makers. Although they already have an access token which has not expired yet, they repeat to discard such a valid access token and request a new one. If this happens, access tokens which are not used but cannot be deleted (because they have not expired yet) are accumulated in the database.</p>
<p>To prevent this situation, save the timestamp at which an access token is last used into the database in addition to the timestamp at which the access token will expire, and run a program periodically which deletes access tokens unused long time. Of course, it depends on characteristics of a service whether it is acceptable or not to delete unused access tokens while they have not expired.</p>
<p>Once before, I met an engineer who had worked in a project for OAuth implementation in a certain big company while he belonged to it. He told me that the system had built without consideration about access token deletion and so the database of the system was perhaps holding hundreds of millions of access tokens then. Scary, scary. When a system which generates something is developed, the timing to delete the generated thing should be considered at the same time.</p>
<h2>8.1. Redirect URI Validation</h2>
<p>In May, 2014, a Ph.D. student in Singapore posted an <a rel="noopener nofollow" href="http://tetraph.com/covert_redirect/oauth2_openid_covert_redirect.html" target="_blank">article</a>, and it made people buzz about <em>&ldquo;Vulnerability in OAuth?&rdquo;</em> It is an issue about so-called Covert Redirect. Those who understand OAuth 2.0 correctly soon realized that it was not due to vulnerability in the specification but just due to the improper implementations. However, the topic made so many people upset that experts in the OAuth area could not help writing explanatory documents. <em>&ldquo;</em><a rel="noopener nofollow" href="http://www.thread-safe.com/2014/05/covert-redirect-and-its-real-impact-on.html" target="_blank"><em>Covert Redirect and its real impact on OAuth and OpenID Connect</em></a><em>&rdquo;</em> by Mr. John Bradley is one of such documents.</p>
<p>If redirect URIs are not treated properly, security problems occur. How redirect URIs should be treated are described in the related specifications, but it is difficult to implement it correctly because there are many things to care about, for example, (a) requirements by RFC 6749 and those by OpenID Connect are different and (b) the value of <code>application_type</code> attribute of client applications must be taken into consideration.</p>
<p>How correctly the part treating redirect URIs is implemented depends on how carefully and exhaustively the implementor has perused the related specifications. Therefore, reading the implementation code of the part can make a good guess about the implementation quality of the entire authorization server. So, everyone, make your best efforts to implement it!</p>
<p>&hellip;&hellip; I would feel sorry if I coldly deserted you like this who have read my long post so far, so <strong>I show you Authlete&rsquo;s implementation know-how</strong>. The following is pseudo code to handle the <code>redirect_uri</code> parameter contained in an authorization request. Note that the pseudo code dare to be written without being split into methods for browsability, but <strong>in the actual Authlete&rsquo;s implementation, the code flow is finely split into methods. Therefore, and for performance purposes, the actual code flow is different from the pseudo code.</strong> (It would be shame if the actual implementation contained so many nested if&rsquo;s and for&rsquo;s like the pseudo code.)</p>
<pre><span>// Extract the value of the 'redirect_uri' parameter from<br>// the authorization request.<br>redirectUri = ...</span><span>// Remember whether a redirect URI was explicitly given.<br>// It must be checked later in the implementation of the<br>// token endpoint because RFC 6749 states as follows.<br>//<br>// redirect_uri<br>// REQUIRED, if the "redirect_uri" parameter was<br>// included in the authorization request as described<br>// in Section 4.1.1, and their values MUST be identical.<br>//<br>explicit = (redirectUri != null);</span><span>// Extract registered redirect URIs from the database.<br>registeredRedirectUris = ...</span><span>// Requirements by RFC 6749 (OAuth 2.0) and those by<br>// OpenID Connect are different. Therefore, the code flow<br>// branches according to whether the request is an OpenID<br>// Connect request or not. This is judged by whether the<br>// 'scope' request parameter contains 'openid' as a value.<br>if ( 'openid' is included in 'scope' )<br>{<br> // Check requirements by OpenID Connect.</span><span> // If the 'redirect_uri' is not contained in the request.<br> if ( redirectUri == null )<br> {<br> // The 'redirect_uri' parameter is mandatory in<br> // OpenID Connect. It's optional in RFC 6749.<br> throw new Exception(<br> "The 'redirect_uri' parameter is missing.");<br> }</span><span> // For each registered redirect URI.<br> for ( registeredRedirectUri : registeredRedirectUris )<br> {<br> // 'Simple String Comparison' is required by the<br> // specification.<br> if ( registeredRedirectUri.equals( redirectUri ) )<br> {<br> // OK. The redirect URI specified by the<br> // authorization request is registered.<br> registered = true;<br> break;<br> }<br> }</span><span> // If the redirect URI specified by the authorization<br> // request matches none of the registered redirect URIs.<br> if ( registered == false )<br> {<br> throw new Exception(<br> "The redirect URI is not registered.");<br> }<br>}<br>else<br>{<br> // Check requirements by RFC 6749.</span><span> // If redirect URIs are not registered at all.<br> if ( registeredRedirectUris.size() == 0 )<br> {<br> // RFC 6749, 3.1.2.2. Registration Requirements says<br> // as follows:<br> //<br> // The authorization server MUST require the<br> // following clients to register their<br> // redirection endpoint:<br> //<br> // o Public clients.<br> // o Confidential clients utilizing the<br> // implicit grant type.</span><span> // If the type of the client application which made<br> // the authorization request is 'public'.<br> if ( client.getClientType() == PUBLIC )<br> {<br> throw new Exception(<br> "A redirect URI must be registered.");<br> }<br> // If the client type is 'confidential' and if the<br> // authorization flow is 'Implicit Flow'. If the<br> // 'response_type' request parameter contains either<br> // or both of 'token' and 'id_token', the flow should<br> // be treated as a kind of 'Implicit Flow'.<br> else if ( responseType.requiresImplicitFlow() )<br> {<br> throw new Exception(<br> "A redirect URI must be registered.");<br> }<br> }</span><span> // If the authorization request does not contain the<br> // 'redirect_uri' request parameter.<br> if ( redirectUri == null )<br> {<br> // If redirect URIs are not registered at all,<br> // or if multiple redirect URIs are registered.<br> if ( registeredRedirectUris.size() != 1 )<br> {<br> // A redirect URI must be explicitly specified<br> // by the 'redirect_uri' parameter.<br> throw new Exception(<br> "The 'redirect_uri' parameter is missing.");<br> }</span><span> // One redirect URI is registered. Use it as the<br> // default value of redirect URI.<br> redirectUri = registeredRedirectUris[0];<br> }<br> // The authorization request contains the 'redirect_uri'<br> // parameter, but redirect URIs are not registered.<br> else if ( registeredRedirectUris.size() == 0 )<br> {<br> // The code flow reaches here if and only if the<br> // client type is 'confidential' and the authorization<br> // flow is not 'Implicit Flow'. In this case, the<br> // redirect URI specified by the 'redirect_uri'<br> // parameter of the authorization request is used<br> // although it is not registered. However,<br> // requirements written in RFC 6749, 3.1.2.<br> // Redirection Endpoint are checked.</span><span> // If the specified redirect URI is not an absolute one.<br> if ( redirectUri.isAbsolute() == false )<br> {<br> throw new Exception(<br> "The 'redirect_uri' is not an absolute URI.");<br> }</span><span> // If the specified redirect URI has a fragment part.<br> if ( redirectUri.getFragment() != null )<br> {<br> throw new Exception(<br> "The 'redirect_uri' has a fragment part.");<br> }<br> }<br> else<br> {<br> // If the specified redirect URI is not an absolute one.<br> if ( redirectUri.isAbsolute() == false )<br> {<br> throw new Exception(<br> "The 'redirect_uri' is not an absolute URI.");<br> }</span><span> // If the specified redirect URI has a fragment part.<br> if ( redirectUri.getFragment() != null )<br> {<br> throw new Exception(<br> "The 'redirect_uri' has a fragment part.");<br> }</span><span> // For each registered redirect URI.<br> for (registeredRedirectUri : registeredRedirectUris )<br> {<br> // If the registered redirect URI is a full URI.<br> if ( registeredRedirectUri.getQuery() != null )<br> {<br> // 'Simple String Comparison'<br> if ( registeredRedirectUri.equals( redirectUri ) )<br> {<br> // The specified redirect URI is registered.<br> registered = true;<br> break;<br> }</span><span> // This registered redirect URI does not match.<br> continue;<br> }</span><span> // Compare the scheme parts.<br> if ( registeredRedirectUri.getScheme().equals(<br> redirectUri.getScheme() ) == false )<br> {<br> // This registered redirect URI does not match.<br> continue;<br> }</span><span> // Compare the user information parts. Here I use<br> // an imaginary method 'equalsSafely()' because<br> // the code would become too long if I inlined it.<br> // The method compares arguments without throwing<br> // any exception even if either or both of the<br> // arguments are null.<br> if ( equalsSafely(<br> registeredRedirectUri.getUserInfo(),<br> redirectUri.getUserInfo() ) == false )<br> {<br> // This registered redirect URI does not match.<br> continue;<br> }</span><span> // Compare the host parts. Ignore case sensitivity.<br> if ( registeredRedirectUri.getHost().equalsIgnoreCase(<br> redirectUri.getHost() ) == false )<br> {<br> // This registered redirect URI does not match.<br> continue;<br> }</span><span> // Compare the port parts. Here I use an imaginary<br> // method 'getPortOrDefaultPort()' because the<br> // code would become too long if I inlined it. The<br> // method returns the default port number of the<br> // scheme when 'getPort()' returns -1. The last<br> // resort is 'URI.toURL().getDefaultPort()'. -1 is<br> // returned If 'getDefaultPort()' throws an exception.<br> if ( getPortOrDefaultPort( registeredRedirectUri ) !=<br> getPortOrDefaultPort( redirectUri ) )<br> {<br> // This registered redirect URI does not match.<br> continue;<br> }</span><span> // Compare the path parts. Here I use the imaginary<br> // method 'equalsSafely()' again.<br> if ( equalsSafely( registeredRedirectUri.getPath(),<br> redirectUri.getPath() ) == false )<br> {<br> // This registered redirect URI does not match.<br> continue;<br> }</span><span> // The specified redirect URI is registered.<br> registered = true;<br> break;<br> }</span><span> // If none of the registered redirect URI matches.<br> if ( registered == false )<br> {<br> throw new Exception(<br> "The redirect URI is not registered.");<br> }<br> }<br>}</span><span>// Check requirements by the 'application_type' of the client.</span><span>// If the value of the 'application_type' attribute is 'web'.<br>if ( client.getApplicationType() == WEB )<br>{<br> // If the authorization flow is 'Implicit Flow'. When the<br> // 'response_type' request parameter of the authorization<br> // request contains either or both of 'token' and 'id_token',<br> // it should be treated as a kind of 'Implicit Flow'.<br> if ( responseType.requiresImplicitFlow() )<br> {<br> // If the scheme of the redirect URI is not 'https'.<br> if ( "https".equals( redirectUri.getScheme() ) == false )<br> {<br> // The scheme part of the redirect URI must be <br> // 'https' when a client application whose<br> // 'application_type' is 'web' uses 'Implicit Flow'.<br> throw new Exception(<br> "The scheme of the redirect URI is not 'https'.");<br> }</span><span> // If the host of the redirect URI is 'localhost'.<br> if ( "localhost".equals( redirectUri.getHost() ) )<br> {<br> // The host of the redirect URI must not be<br> // 'localhost' when a client application whose<br> // 'application_type' is 'web' uses 'Implicit Flow'.<br> throw new Exception(<br> "The host of the redirect URI is 'localhost'.");<br> }<br> }<br>}<br>// If the value of the 'application_type' attribute is 'native'.<br>else if ( client.getApplicationType() == NATIVE )<br>{<br> // If the scheme of the redirect URI is 'https'.<br> if ( "https".equals( redirectUri.getScheme() ) )<br> {<br> // The scheme of the redirect URI must not be 'https'<br> // when the 'application_type' of the client is 'native'.<br> throw new Exception(<br> "The scheme of the redirect URI is 'https'.");<br> }</span><span> // If the scheme of the redirect URI is 'http'.<br> if ( "http".equals( redirectUri.getScheme() ) )<br> {<br> // If the host of the redirect URI is not 'localhost'.<br> if ( "localhost".equals(<br> redirectUri.getHost() ) == false )<br> {<br> // When a client application whose 'application_type'<br> // is 'native' uses a redirect URI whose scheme is<br> // 'http', the host port of the URI must be<br> // 'localhost'.<br> throw new Exception(<br> "The host of the redirect URI is not 'localhost'.");<br> }<br> }<br>}<br>// If the value of the 'application_type' attribute is neither<br>// 'web' or 'native'.<br>else<br>{<br> // As mentioned above, Authlete allows 'unspecified' as a<br> // value of the 'application_type' attribute. Therefore,<br> // no exception is thrown here.<br>}</span></pre>
<h2>8.2. Other&rsquo;s implementation</h2>
<p>In OpenID Connect, the <code>redirect_uri</code> parameter is mandatory and requirements about how to check whether a presented redirect URI is registered or not are just &lsquo;Simple String Comparison&rsquo;. Therefore, if what you have to care about is OpenID Connect only, implementations will be simple. For example, in <a rel="noopener nofollow" href="https://github.com/IdentityServer/IdentityServer3" target="_blank"><strong>IdentityServer3</strong></a> which has won about 1,700 stars on GitHub as of October, 2016 and been certified by <a rel="noopener nofollow" href="http://openid.net/certification/" target="_blank"><strong>OpenID Certification</strong></a> program, checking a redirect URI is implemented as follows (excerpt from <code><a rel="noopener nofollow" href="https://github.com/IdentityServer/IdentityServer3/blob/master/source/Core/Services/Default/DefaultRedirectUriValidator.cs" target="_blank">DefaultRedirectUriValidator.cs</a></code> with additional newlines for formatting).</p>
<pre><span>public virtual Task&lt;bool&gt; IsRedirectUriValidAsync(<br> string requestedUri, Client client)<br>{<br> return Task.FromResult(<br> StringCollectionContainsString(<br> client.RedirectUris, requestedUri));<br>}</span></pre>
<p>That OpenID Connect only is cared about means, to put it the other way around, that traditional authorization code flow and implicit flow which do not contain <code>openid</code> in the <code>scope</code> request parameter are NOT accepted by the authorization server. That is, such an authorization server cannot respond to any existing OAuth 2.0 client applications.</p>
<p>So, does IdentityServer3 reject traditional authorization requests? Take a look at <code><a rel="noopener nofollow" href="https://github.com/IdentityServer/IdentityServer3/blob/master/source/Core/Validation/AuthorizeRequestValidator.cs" target="_blank">AuthorizeRequestValidator.cs</a></code> and you will find this (formatting is adjusted):</p>
<pre><span>if (request.RequestedScopes.Contains(<br> Constants.StandardScopes.OpenId))<br>{<br> request.IsOpenIdRequest = true;<br>}<br>//////////////////////////////////////////////////////////<br><br><br>// check scope vs response_type plausability<br>//////////////////////////////////////////////////////////<br>var requirement =<br>Constants.ResponseTypeToScopeRequirement[request.ResponseType];<br>if (requirement == Constants.ScopeRequirement.Identity ||<br> requirement == Constants.ScopeRequirement.IdentityOnly)<br>{<br> if (request.IsOpenIdRequest == false)<br> {<br> LogError("response_type requires the openid scope", request);<br> return Invalid(request, ErrorTypes.Client);<br> }<br>}</span></pre>
<p>You don&rsquo;t have to understand details of this code. The point is that there are paths which allow cases where <code>openid</code> is not contained in the <code>scope</code> parameter. That is, traditional authorization requests are accepted. If so, the implementation of IdentityServer3 is not correct. However, on the other hand, at another place in <code>AuthorizeRequestValidator.cs</code>, the implementation rejects all authorization requests that do not contain the <code>redirect_uri</code> parameter like below (formatting is adjusted).</p>
<pre><span>//////////////////////////////////////////////////////////<br>// redirect_uri must be present, and a valid uri<br>//////////////////////////////////////////////////////////<br>var redirectUri = request.Raw.Get(Constants.AuthorizeRequest.RedirectUri);<br>if (redirectUri.IsMissingOrTooLong(<br><br><br> _options.InputLengthRestrictions.RedirectUri))<br>{<br> LogError("redirect_uri is missing or too long", request);<br> return Invalid(request);<br>}</span></pre>
<p>So, the implementation does not have to care about the case where the <code>redirect_uri</code> parameter is omitted. But, because the <code>redirect_uri</code> parameter is optional in RFC 6749, the behavior &mdash; authorization requests without the <code>redirect_uri</code> parameter are unconditionally rejected although traditional authorization requests are accepted &mdash; is a violation of the specification. In addition, IdentityServer3 does not do validation for the <code>application_type</code> attribute. To implement the validation, as a first step, a property for the <code>application_type</code> attribute must be added to the model class that represents a client application ( <code><a rel="noopener nofollow" href="https://github.com/IdentityServer/IdentityServer3/blob/master/source/Core/Models/Client.cs" target="_blank">Client.cs</a></code>) as the current implementation misses it.</p>
<p>Subtle violations of the specifications are sometimes called &ldquo;dialects&rdquo;. The word &ldquo;dialect&rdquo; may give an impression of &ldquo;acceptable&rdquo;, but violations are violations. If there are no dialects, it will be enough to have one generic OAuth 2.0 / OpenID Connect library for each computer language. But, in the real world, custom client libraries are needed for authorization servers which violate the specifications.</p>
<p>The reason Facebook&rsquo;s OAuth flow requires its custom client library is that there are many violations of the specifications in Facebook&rsquo;s OAuth implementation. For example, (1) commas are used as a delimiter of scope list (it should be spaces), (2) the format of the response from the token endpoint is <code>application/x-www-form-urlencoded</code> (it should be JSON), and (3) the name of the parameter for access token&rsquo;s expiration date is <code>expires</code> (it should be <code>expires_in</code>).</p>
<p>Not only Facebook but also other big names have violations of the specifications. The following are other examples.</p>
<h2>9.1. Delimiter of Scope List</h2>
<p>Scope names are listed in the <code>scope</code> parameter of requests to an <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc6749#section-3.1" target="_blank">authorization endpoint</a> and a <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc6749#section-3.2" target="_blank">token endpoint</a>. RFC 6749, <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc6749#section-3.3" target="_blank">3.3. Access Token Scope</a> requires that spaces be used as delimiters, but the following OAuth implementations use commas:</p>
<ul>
<li>Facebook</li>
<li>GitHub</li>
<li>Spotify</li>
<li>Discus</li>
<li>Todoist</li>
</ul>
<h2>9.2. Response Format of Token Endpoint</h2>
<p>RFC 6749, <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc6749#section-5.1" target="_blank">5.1. Successful Response</a> requires that the format of a successful response from a token endpoint be JSON, but the following OAuth implementations use <code>application/x-www-form-urlencoded</code>:</p>
<ul>
<li>Facebook</li>
<li>Bitly</li>
<li>GitHub</li>
</ul>
<p>The default format is <code>application/x-www-form-urlencoded</code>, but GitHub provides a means to request JSON.</p>
<h2>9.3. token_type in Response from Token Endpoint</h2>
<p>RFC 6749, <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc6749#section-5.1" target="_blank">5.1. Successful Response</a> requires that the <code>token_type</code> parameter be included in a successful response from a token endpoint, but the following OAuth implementation does not include it:</p>
<ul>
<li>Slack</li>
</ul>
<p>Salesforce once had this issue, too (<a rel="noopener nofollow" href="https://developer.salesforce.com/forums/?id=906F00000009BCRIA2" target="_blank">OAuth Access Token Response Missing token_type</a>), but it has been fixed.</p>
<h2>9.4. token_type Inconsistency</h2>
<p>The following OAuth implementation claims that the token type is &ldquo;Bearer&rdquo;, but its resource endpoints do not accept an access token by the means defined in <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc6750" target="_blank">RFC 6750</a> (The OAuth 2.0 Authorization Framework: Bearer Token Usage):</p>
<ul>
<li>GitHub (it accepts an access token via the format of <code>Authorization: token OAUTH-TOKEN</code>)</li>
</ul>
<h2>9.5. grant_type Is Not Required</h2>
<p>The <code>grant_type</code> parameter is mandatory at a token endpoint, but the following OAuth implementations don&rsquo;t require it:</p>
<ul>
<li>GitHub</li>
<li>Slack</li>
<li>Todoist</li>
</ul>
<h2>9.6. Unofficial Values for The error Parameter</h2>
<p>The specifications have defined some values for the <code>error</code> parameter which is included in an error response from an authorization server, but the following OAuth implementations define their own:</p>
<ul>
<li>GitHub (e.g. <code>application_suspended</code>)</li>
<li>Todoist (e.g. <code>bad_authorization_code</code>)</li>
</ul>
<h2>9.7. Bad Parameter Name on Error</h2>
<p>The following OAuth implementation uses <code>errorCode</code> instead of <code>error</code> when it returns an error code:</p>
<ul>
<li>LINE</li>
</ul>
<h2>10.1. PKCE Is A MUST</h2>
<p>Do you know <strong>PKCE</strong>? It is a specification defined as <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc7636" target="_blank">RFC 7636</a> (Proof Key for Code Exchange by OAuth Public Clients) and was published in September, 2015. It is a countermeasure against the <strong>authorization code interception attack</strong>.</p>
<figure>
<div>
<div>
<div>
<div><img width="691" src="https://miro.medium.com/max/60/1*n0kpaYqPLb_yvGYM5uPO9g.png?q=20" height="435"></div>
<img width="691" height="435">
</div>
</div>
</div>
</figure>
<p>Some conditions are required for the attack to succeed, but if you are thinking of releasing a smartphone application, it is strongly recommended that PKCE be supported by both the client application and the authorization server. Otherwise, a malicious application may intercept an authorization code issued by the authorization server and exchange it with a valid access token at the token endpoint of the authorization server.</p>
<p>It is in October, 2012 that RFC 6749 (The OAuth 2.0 Authorization Framework) was released, so <strong>even developers who are familiar with OAuth 2.0 may not know RFC 7636</strong> which was released relatively recently in September, 2015. However, it should be noted that <a rel="noopener nofollow" href="https://tools.ietf.org/html/draft-ietf-oauth-native-apps" target="_blank">the draft of &ldquo;OAuth 2.0 for Native Apps&rdquo;</a> states its support is a MUST under some conditions.</p>
<blockquote>
<p><strong>Both the client and the Authorization Server MUST support PKCE [RFC7636] to use custom URI schemes, or loopback IP redirects.</strong> Authorization Servers SHOULD reject authorization requests using a custom scheme, or loopback IP as part of the redirection URI if the required PKCE parameters are not present, returning the error message as defined in Section 4.4.1 of PKCE [RFC7636]. It is RECOMMENDED to use PKCE [RFC7636] for app-claimed HTTPS redirect URIs, even though these are not generally subject to interception, to protect against attacks on inter-app communication.</p>
</blockquote>
<p>The authorization endpoint of an authorization server that supports RFC 7636 accepts two request parameters, <code>code_challenge</code> and <code>code_challenge_method</code>, and the token endpoint accepts <code>code_verifier</code>. And in the implementation of the token endpoint, the authorization server computes the value of code challenge using (a) the code verifier presented by the client application and (b) the code challenge method specified by the client application at the authorization endpoint. If the computed code challenge and the value of the <code>code_challenge</code> parameter presented by the client application at the authorization endpoint are equal, it can be said that the entity that made the authorization request and the entity that made the token request are identical. Thus, an authorization server can avoid issuing an access token to a malicious application which is different from the entity that made the authorization request.</p>
<figure>
<div>
<div>
<div>
<div><img width="525" src="https://miro.medium.com/max/44/1*cc_j8ia_MSBHWbXRDSDY6A.png?q=20" height="701"></div>
<img width="525" height="701">
</div>
</div>
</div>
</figure>
<p>The entire flow of RFC 7636 is explained with illustration at Authlete&rsquo;s website: <a rel="noopener nofollow" href="https://www.authlete.com/documents/article/pkce/index" target="_blank"><em>Proof Key for Code Exchange (RFC 7636)</em></a>. If you are interested, please read it.</p>
<h2>10.2. Server-Side Implementation</h2>
<p>In the implementation of an authorization endpoint, what an authorization server has to do is to save the values of the <code>code_challenge</code> parameter and the <code>code_challenge_method</code> parameter contained in an authorization request into the database. So, there is nothing interesting in the implementation code. Something to note is just that <strong>an authorization server which wants to support PKCE has to add columns for</strong> <code><strong>code_challenge</strong></code> <strong>and</strong> <code><strong>code_challenge_method</strong></code> <strong>into the database table storing authorization codes</strong>.</p>
<p>The entire source code of Authlete is confidential, but for your interest, here I show you the actual Authlete&rsquo;s implementation that validates the value of the <code>code_verifier</code> parameter at the token endpoint.</p>
<pre><span>private void validatePKCE(AuthorizationCodeEntity acEntity)<br>{<br> // See RFC 7636 (Proof Key for Code Exchange) for details.<br> // Get the value of 'code_challenge' which was contained in<br><br><br> // the authorization request.<br> String challenge = acEntity.getCodeChallenge();<br> if (challenge == null)<br><br><br> {<br> // The authorization request did not contain<br> // 'code_challenge'.<br> return;<br> }<br> // If the authorization request contained 'code_challenge',<br><br><br> // the token request must contain 'code_verifier'. Extract<br> // the value of 'code_verifier' from the token request.<br> String verifier = extractFromParameters(<br> "code_verifier", invalid_grant, A050312, A050313, A050314);<br> // Compute the challenge using the verifier<br><br><br> String computedChallenge = computeChallenge(acEntity, verifier);<br> if (challenge.equals(computedChallenge))<br><br><br> {<br> // OK. The presented code_verifier is valid.<br> return;<br> }<br> // The code challenge value computed with 'code_verifier'<br><br><br> // is different from 'code_challenge' contained in the<br> // authorization request.<br> throw toException(invalid_grant, A050315);<br>}<br>private String computeChallenge(<br><br><br> AuthorizationCodeEntity acEntity, String verifier)<br>{<br> CodeChallengeMethod method = acEntity.getCodeChallengeMethod();<br> // This should not happen, but just in case.<br><br><br> if (method == null)<br> {<br> // Use 'plain' as the default value required by RFC 7636.<br> method = CodeChallengeMethod.PLAIN;<br> }<br> switch (method)<br><br><br> {<br> case PLAIN:<br> // code_verifier<br> return verifier;<br> case S256:<br><br><br> // BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))<br> return computeChallengeS256(verifier);<br> default:<br><br><br> // The value of code_challenge_method extracted<br> // from the database is not supported.<br> throw toException(server_error, A050102);<br> }<br>}<br>private String computeChallengeS256(String verifier)<br><br><br>{<br> // BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))<br> // SHA256<br><br><br> byte[] hash =<br> Digest.getInstanceSHA256().update(verifier).digest();<br> // BASE64URL<br><br><br> return SecurityUtils.encode(hash);<br>}</span></pre>
<p>The <code><a rel="noopener nofollow" href="http://takahikokawasaki.github.io/nv-digest/com/neovisionaries/security/Digest.html" target="_blank">Digest</a></code> class used in the implementation of <code>computeChallengeS256(String)</code> method is included in my open-source library, <a rel="noopener nofollow" href="https://github.com/TakahikoKawasaki/nv-digest" target="_blank"><strong>nv-digest</strong></a>. It is a utility library to make digest computation easy. With this library, computing an SHA-256 digest value can be written in a line as shown below.</p>
<pre><span>byte[] hash = Digest.getInstanceSHA256().update(verifier).digest();</span></pre>
<h2>10.3. Client-Side Implementation</h2>
<p>What a client application has to do for PKCE are two. One is to generate a random code verifier which consists of 43&ndash;128 letters, compute the code challenge using the code verifier and the code challenge method (<code>plain</code> or <code>S256</code>), and include the computed code challenge and the code challenge method as the values of the <code>code_challenge</code> parameter and the <code>code_challenge_method</code> parameter in an authorization request. The other is to include the code verifier in a token request.</p>
<p>As examples of client-side implementations, I introduce the following two.</p>
<ol>
<li><a rel="noopener nofollow" href="http://openid.github.io/AppAuth-Android/" target="_blank"><strong>AppAuth for Android</strong></a></li>
<li><a rel="noopener nofollow" href="https://github.com/openid/AppAuth-iOS" target="_blank"><strong>AppAuth for iOS</strong></a></li>
</ol>
<p>They are SDKs to communicate with an OAuth 2.0 and OpenID Connect server. They claim that they include best practices and support PKCE.</p>
<p>If you implement a computation logic for <code>code_challenge_method=S256</code>, you can test it by checking whether the value of the code challenge becomes <code>E9Melhoa2OwvFrEMTJguCHaoeK1t8URWbuGJSstw-cM</code> when the value of the code verifier is <code>dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk</code>. These values are found as example values in &ldquo;<a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc7636#appendix-B" target="_blank">Appendix B. Example for the S256 code_challenge_method</a>&rdquo; of <a rel="noopener nofollow" href="https://tools.ietf.org/html/rfc7636" target="_blank">RFC 7636</a>.</p>
<p>Some may say it is easy to implement OAuth and OpenID Connect, and others may say it&rsquo;s not. In either case, as a matter of fact, even big tech companies such as Facebook and GitHub that have sufficient budget and human resources have failed to implement OAuth and OpenID Connect correctly. Famous open-source projects such as Apache Oltu and Spring Security have problems, too. Therefore, if you implement OAuth and OpenID Connect by yourself, take it seriously and prepare a decent development team. Otherwise, security risks would be increased.</p>
<p>It is not hard to implement RFC 6749 only, but it would drive you crazy to implement OpenID Connect from scratch. So, it is recommended to use an existing implementation as a starting point. The first step will be to search the page &ldquo;<a rel="noopener nofollow" href="http://openid.net/developers/libraries/" target="_blank">Libraries, Products, and Tools</a>&rdquo; in the OpenID Connect website for software related to OAuth and OpenID Connect (although Authlete is not listed there). Of course, as a co-founder of <a rel="noopener nofollow" href="https://www.authlete.com/" target="_blank">Authlete, Inc.</a>, I will be glad if you choose Authlete.</p>
<p>Thank you for reading this long post.</p>
</div>
</div>
</section>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Medium https://medium.com<br>
<br>
May 27, 2020 at 11:09AM
